rust/src/nasl/builtin/network/socket.rs

```diff
@@ -512,6 +548,91 @@ impl NaslSockets {
         Ok(NaslValue::Number(fd as i64))
     }
 
+    fn open_priv_sock(
+        &self,
+        addr: IpAddr,
+        dport: i64,
+        sport: Option<i64>,
+        tcp: bool,
+    ) -> Result<NaslValue, FunctionErrorKind> {
+        let dport = verify_port(dport)?;
+
+        if let Some(sport) = sport {
+            let sport = verify_port(sport)?;
+            let fd = if tcp {
+                // TODO: set timeout to global recv timeout when available
+                let timeout = Duration::from_secs(10);
+                self.wait_before_next_probe();
+                let tcp = TcpConnection::connect_priv(addr, sport, dport, timeout)?;
+                self.add(NaslSocket::Tcp(Box::new(tcp)))
+            } else {
+                let udp = UdpConnection::new_priv(addr, sport, dport)?;
+                self.add(NaslSocket::Udp(udp))
+            };
+            return Ok(NaslValue::Number(fd as i64));
+        }
+
+        for sport in (1..=1023).rev() {
+            let fd = if tcp {
+                // TODO: set timeout to global recv timeout when available
+                let timeout = Duration::from_secs(10);
+                self.wait_before_next_probe();
+                if let Ok(tcp) = TcpConnection::connect_priv(addr, sport, dport, timeout) {
+                    self.add(NaslSocket::Tcp(Box::new(tcp)))
+                } else {
+                    continue;
+                }
+            } else {
+                if let Ok(udp) = UdpConnection::new_priv(addr, sport, dport) {
+                    self.add(NaslSocket::Udp(udp))
+                } else {
+                    continue;
+                }
+            };
+            return Ok(NaslValue::Number(fd as i64));
+        }
```

Comment:
I would find this a lot easier to read with something like this (couldn't think of a name for the first method):

```rust
    fn open_priv_sock_internal(&self, addr: IpAddr, sport: u16, dport: u16, tcp: bool) -> Result<usize, FunctionErrorKind> {
        if tcp {
            // TODO: set timeout to global recv timeout when available
            let timeout = Duration::from_secs(10);
            self.wait_before_next_probe();
            let tcp = TcpConnection::connect_priv(addr, sport, dport, timeout)?;
            Ok(self.add(NaslSocket::Tcp(Box::new(tcp))))
        } else {
            let udp = UdpConnection::new_priv(addr, sport, dport)?;
            Ok(self.add(NaslSocket::Udp(udp)))
        }
    }

    fn open_priv_sock(
        &self,
        addr: IpAddr,
        dport: i64,
        sport: Option<i64>,
        tcp: bool,
    ) -> Result<usize, FunctionErrorKind> {
        let dport = verify_port(dport)?;

        if let Some(sport) = sport {
            let sport = verify_port(sport)?;
            return Ok(self.open_priv_sock_internal(addr, sport, dport, tcp)?);
        }

        for sport in (1..=1023).rev() {
            let res = self.open_priv_sock_internal(addr, sport, dport, tcp);
            if let Ok(fd) = res {
                return Ok(fd)
            }
        }
        Err(FunctionErrorKind::Diagnostic(
            format!(
                "Unable to open priv socket to {} on any socket from 1-1023",
                addr
            ),
            None,
        ))
    }
```